atef/plan_utils.py

```diff
@@ -0,0 +1,242 @@
+
+        return cls.instance
+
+    def run_plan(
+        self,
+        state: BlueskyState,
+        item: Dict[str, Any],
+        identifier: str
+    ) -> Tuple[UUID, ...]:
+        """
+        Run a plan item, and store the resulting UUIDs in
+        ``state`` under ``identifier``
+        """
+        # separate device from component attributes
+        base, cpt = separate_attrs_from_strings(item)
+
+        # fill device, plan names with objects from namespace
+        parsed_plan = prepare_plan(
+            base,
+            plans_in_nspace=state.plans_in_ns,
+            devices_in_nspace=state.devices_in_ns,
+            allowed_plans=state.allowed_plans,
+            allowed_devices=state.allowed_devices
+        )
+
+        # reconstruct args and kwargs (operator.attrgetter(attr)(device))
+        parsed_args = get_attr_from_strings(parsed_plan['args'], cpt['args'])
+        parsed_kwargs = get_attr_from_strings(parsed_plan['kwargs'], cpt['kwargs'])
+
+        # actually run plan and store uuids
+        logger.debug(f'running plan: {parsed_plan["callable"]}'
+                     f'(*{parsed_args}, **{parsed_kwargs})')
+        run_uuids = self.RE(parsed_plan["callable"](
+                            *parsed_args, **parsed_kwargs))
+        state.run_map[identifier] = run_uuids
+        return run_uuids
+
+
+def separate_attrs_from_strings(item: Any) -> Tuple[Any, Any]:
+    """ Recursively strip attributes from strings, return other items unchanged """
+    if isinstance(item, dict):
+        base, attr = {}, {}
+        for key, value in item.items():
+            device_val, cpt_val = separate_attrs_from_strings(value)
+            base[key] = device_val
+            attr[key] = cpt_val
+        return base, attr
+    elif isinstance(item, list):
+        base, attr = [], []
+        for value in item:
+            base_val, attr_val = separate_attrs_from_strings(value)
+            base.append(base_val)
+            attr.append(attr_val)
+        return base, attr
+    elif isinstance(item, str):
+        try:
+            device, cpt = item.split('.', maxsplit=1)
+        except ValueError:
+            return item, item
+        return device, cpt
+    else:
+        return item, item
+
+
+def get_attr_from_strings(base: Any, cpt: Any) -> Any:
+    """
+    Get ophyd copmonents from the devices in ``base``, if there is an attribute
+    in ``cpt``.  Traverses ``base`` and ``cpt`` in parallel
+    """
+    if base == cpt:
+        return base
+
+    if isinstance(base, dict):
+        return {k: get_attr_from_strings(base[k], cpt[k]) for k in base.keys()}
+    elif isinstance(base, list):
+        return [get_attr_from_strings(b, c) for b, c in zip_longest(base, cpt)]
+    elif isinstance(base, ophyd.Device) and (base.name != cpt):
+        return operator.attrgetter(cpt)(base)
+    else:
+        return base
+
+
+def register_run_identifier(state: BlueskyState, name: str) -> str:
```

Comment:
This works on `BlueskyState`, so maybe it should be a method there

The 3rd attempt for registering a name seems like it'd be `new_name_1_2_3` - is that intentional?
Might just want to create a hash, throw it on the end of the "user-defined name" and be done with it

Reply:
erm, it's supposed to be `new_name_1` -> `new_name_2`, whoops.  I wanted these to be human readable since users would have to choose from a dropdown